### PR TITLE
add nodePackages vega-cli and vega-lite

### PR DESCRIPTION
**Pull request by MMesch**
###### Motivation for this change

Adds the [vega](https://vega.github.io/vega/usage/#cli) and the [vega-lite](https://vega.github.io/vega-lite/usage/compile.html#cli) command line tools for plotting. This means the binaries:

```
vg2svg
vg2pdf
vg2png
vl2vg
vl2pdf
vl2svg
vl2png
```

Vegalite required substituting some lines that call `npx` to install `vega` on the fly. Maybe there is a better solution. I tested these tools with:

```
nix-shell --pure -I nixpkgs=./nixpkgs -p busybox -p nodePackages.vega-cli --command "wget -O - https://vega.github.io/vega/examples/bar-chart.vg.json | vg2png > vega-cli-test.png"
```

```
nix-shell --pure -I nixpkgs=./nixpkgs -p busybox -p nodePackages.vega-lite --command "wget -O - https://raw.githubusercontent.com/vega/vega-lite/master/examples/specs/bar.vl.json | vl2png > vega-lite-test.png"
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by MMesch** [comment]
I tried to run `nixpkgs-review pr --post-result 97908` and stopped it after 2h because I needed the CPU. It somehow seems to hang on `twemoji`, `mojione` and `sagedoc`. `inkscape` also somehow appears there. The output that I got until then is:

```
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/97908/head:refs/nixpkgs-review/1
$ git worktree add /home/matto/.cache/nixpkgs-review/pr-97908-1/nixpkgs 24fa210e772351f118e5f56e5acbe5eaea38a95f
Preparing worktree (detached HEAD 24fa210e772)
Updating files: 100% (22686/22686), done.
HEAD is now at 24fa210e772 rclone: 1.53.0 -> 1.53.1 (#97953)
$ nix-env -f /home/matto/.cache/nixpkgs-review/pr-97908-1/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit 690fffccf47b4561013912dd37d5bda8ea8484b7
Automatic merge went well; stopped before committing as requested
$ nix-env -f /home/matto/.cache/nixpkgs-review/pr-97908-1/nixpkgs -qaP --xml --out-path --show-trace --meta
39 packages updated:
bitwarden-cli emojione etcher image_optim lumo node_castnow node_create-cycle-app node_imapnotify joplin (1.0.166 → 1.0.167) node_mirakurun netlify-cli (2.61.0 → 2.62.0) node_redoc-cli node_thelounge node_triton node__at_antora_slash_cli node__at_webassemblyjs_slash_wasm-text-gen parity-ui ripcord sage sage sage slack slack twemoji-color-font vimPlugins.coc-css (1.2.4 → 1.2.5) vimPlugins.coc-eslint (1.2.7 → 1.3.0) vimPlugins.coc-git (1.8.3 → 2.0.0) vimPlugins.coc-go (0.10.0 → 0.11.0) vimPlugins.coc-highlight (1.2.5 → 1.2.6) vimPlugins.coc-jest (1.0.3 → 1.0.4) vimPlugins.coc-json (1.2.6 → 1.3.0) vimPlugins.coc-metals (0.9.2 → 0.9.3) vimplugin-coc-prettier vimPlugins.coc-r-lsp (1.1.1 → 1.2.0) vimPlugins.coc-snippets (2.1.28 → 2.2.0) vimPlugins.coc-solargraph (1.1.6 → 1.1.7) vimplugin-coc-stylelint vimPlugins.coc-tsserver (1.5.5 → 1.5.6) vimPlugins.coc-vetur (1.1.12 → 1.1.13)

$ nix build --no-link --keep-going --option build-use-sandbox relaxed -f /home/matto/.cache/nixpkgs-review/pr-97908-1/build.nix
builder for '/nix/store/b42m79dxaz44gjb4gpg3mmlwdr4zxy4h-node_joplin-1.0.167.drv' failed with exit code 1; last 10 log lines:
  npm ERR! errno 1
  npm ERR! keytar@5.6.0 install: `prebuild-install || node-gyp rebuild`
  npm ERR! Exit status 1
  npm ERR! 
  npm ERR! Failed at the keytar@5.6.0 install script.
  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
  
  npm ERR! A complete log of this run can be found in:
  npm ERR!     /build/.npm/_logs/2020-09-14T08_13_48_980Z-debug.log
 [...]
```

**Comment by doronbehar** [comment]
@MMesch It's almost impossible to merge changes to node packages, unfortunately. Could you please fix them and ping me when you are ready? Preferably, `git add` all the files (don't use `--patch` to select only the changes relevant to you). I promise I'll merge it before there are merge conflicts.

Once that'll be done, I'd be happy to get your help on https://github.com/NixOS/nixpkgs/pull/96961 .

**Comment by MMesch** [comment]
Hi @doronbehar 

> It's almost impossible to merge changes to node packages, unfortunately

What do you mean with this? Is it possible or not? I looked on your PR and saw that you suggested generating nix expressions for the packages you want to add independently. I have done that equally for vega-cli and vega-lite on my machine. Do you think it would be better to move forward like this?

Also, I already added `git add` everything! Are there files that you are missing?

**Comment by doronbehar** [comment]
> What do you mean with this?

I mean that as a contributor, with no merge permissions, it's impossible to put out a PR, and when the time comes for a committer to merge it, the PR has no merge conflicts.

> I have done that equally for vega-cli and vega-lite on my machine. Do you think it would be better to move forward like this?

It depends on you. For balena-cli, I realized at a certain point I have to do it because their releases are too frequent for node packages' `generate.sh`. If you'd like to have more influence upon it, I'd suggest to make it have it's own set of generated nix expressions. Plus, for sure you'd probably won't encounter merge conflicts with the generated expressions.

**Comment by doronbehar** [comment]
> Are there files that you are missing?

I don't think so, but I haven't checked.

**Comment by MMesch** [comment]
@doronbehar I have rebased and regenerated node-packages.nix for the two relevant commits. Maybe I didn't understand if you wanted to do anything else. Let me know!

In case there are conflicts again, this can be rebased from the nodePackages folder with:

```
git fetch
git rebase origin/master
git checkout --theirs node-packages.nix
./generate.sh
git add .
git rebase --continue
git checkout --theirs node-packages.nix
./generate.sh
git add .
git rebase --continue
```

(takes ~2h on my machine)

File: pkgs/development/node-packages/default.bak
Comment (doronbehar) [inline pkgs/development/node-packages/default.bak:1]: This file is a Git rebase leftover right?

**Comment by MMesch** [comment]
Right! Sorry. It should be deleted. Do you want me to do it again?

Doron Behar <notifications@github.com> schrieb am Sa., 19. Sep. 2020, 13:41:

File: maintainers/maintainer-list.nix
Comment (doronbehar) [inline maintainers/maintainer-list.nix:5712]: the commit message should be:

```
maintainers: add mmesch
```

**Comment by doronbehar** [comment]
You should be able to spare yourself running `./generate.sh`, but you should need rewriting history a bit.

**Comment by MMesch** [comment]
I hope it's ok now @doronbehar . I ran the `vega-cli` and `vega-lite` tests that I posted above and they work.

File: pkgs/development/node-packages/default.nix
Comment (doronbehar) [inline pkgs/development/node-packages/default.nix:188]:
```suggestion
        # npx tries to install vega from scratch at vegalite runtime if it
        # can't find it. We thus replace it with a direct call to the nix
        # derivation.
        postInstall = ''
```

File: pkgs/development/node-packages/default.nix
Comment (doronbehar) [inline pkgs/development/node-packages/default.nix:195]:
```suggestion
    '';
```

Mixed indentation.

**Review by doronbehar — CHANGES_REQUESTED** [review]

**Comment by MMesch** [comment]
Sorry for the formatting issues. How about now @doronbehar ?

File: pkgs/development/node-packages/default.nix
Comment (doronbehar) [inline pkgs/development/node-packages/default.nix:196]:
```suggestion
          substituteInPlace $out/lib/node_modules/vega-lite/bin/vl2pdf \
            --replace "npx -p vega vg2pdf"  "${self.vega-cli}/bin/vg2pdf"
          substituteInPlace $out/lib/node_modules/vega-lite/bin/vl2svg \
            --replace "npx -p vega vg2svg"  "${self.vega-cli}/bin/vg2svg"
          substituteInPlace $out/lib/node_modules/vega-lite/bin/vl2png \
            --replace "npx -p vega vg2png"  "${self.vega-cli}/bin/vg2png"
      '';
```

Sorry for the nitpick, but it should be perfect now!

**Comment by doronbehar** [comment]
@MMesch after fixing once more the indentation, please apply this patch:

```patch
diff --git i/pkgs/development/node-packages/default.nix w/pkgs/development/node-packages/default.nix
index 71b9bd77411..8d8d9fc89ae 100644
--- i/pkgs/development/node-packages/default.nix
+++ w/pkgs/development/node-packages/default.nix
@@ -181,6 +181,8 @@ let
         # https://sharp.pixelplumbing.com/install
         vips
 
+        libsecret
+        self.node-gyp-build
         self.node-pre-gyp
       ];
     };
```

And commit it as: `joplin: fix build by adding libsecret and node-gyp-build`.

It's not related to your PR, but these are new deps that your update to all of node packages require - since you ran `./generate.sh`.

**Comment by MMesch** [comment]
Ah s... I forgot that one. Unfortunately I can't do it until tomorrow. If
you have the time to amend this and merge go ahead, otherwise it has to
wait.

Thabks for the help and sorry for all the work :( I tried to use nixfmt on
this but it changes a lot of other things.

Doron Behar <notifications@github.com> schrieb am Sa., 19. Sep. 2020, 16:34:

**Comment by MMesch** [comment]
Many thanks for you help @doronbehar . How can I help with your PR?

**Comment by doronbehar** [comment]
TBH I only want someone to approve it for the record, and verify that it works. And comments upon meta data or such would be welcome https://github.com/NixOS/nixpkgs/pull/96961 .

**Pull request merged**
